### ISSUE TITLE
cluster isolation: crashing storaged brings down the coordinator process

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#12595]
### What version of Materialize are you using?

v0.26.1-dev (c5e09ac91)

### How did you install Materialize?

Built from source

### What is the issue?

I was hitting Ctrl+C repeatedly on `UPDATE` queries and got the following:

1. First:
```
2022-05-20T15:21:18.181181Z ERROR mz_orchestrator_process: storage-runtime exited: Ok(ExitStatus(unix_wait_status(9))); relaunching in 5s
```
2. Second:
```
thread 'timely:work-0' panicked at 'called `Result::unwrap()` on an `Err` value: SendError(Register { source_id: (416dc55b-f60e-4a4d-bd48-f3107c73d184, User(6)), worker: 0, request: SourceInstanceRequest { source_id: User(6), dataflow_id: 416dc55b-f60e-4a4d-bd48-f3107c73d184, arguments: SourceInstanceArguments { operators: Some(LinearOperator { predicates: [], projection: [0] }) }, as_of: Antichain { elements: [1653060068221] } }, storage_workers: [0], ok_activator: ArcActivator { inner: Mutex { data: ArcActivatorInner { activated: 0, activators: [], _name: "Dataflow: temp-view-t181-u6-ok-activator", threshold: 1 }, poisoned: false, .. } }, err_activator: ArcActivator { inner: Mutex { data: ArcActivatorInner { activated: 0, activators: [], _name: "Dataflow: temp-view-t181-u6-err-activator", threshold: 1 }, poisoned: false, .. } }, ok_tx: UnboundedSender { chan: Tx { inner: Chan { tx: Tx { block_tail: 0x7fc5957bf600, tail_position: 0 }, semaphore: 0, rx_waker: AtomicWaker, tx_count: 1, rx_fields: "..." } } }, err_tx: UnboundedSender { chan: Tx { inner: Chan { tx: Tx { block_tail: 0x7fc5957bfc00, tail_position: 0 }, semaphore: 0, rx_waker: AtomicWaker, tx_count: 1, rx_fields: "..." } } } })', /home/ubuntu/materialize/src/storage/src/boundary/tcp_boundary.rs:674:45
```

and finally:

```
2022-05-20T15:21:19.524775Z ERROR panic: called `Result::unwrap()` on an `Err` value: ClientError(client is reconnecting)
thread: coordinator
location: /home/ubuntu/materialize/src/coord/src/coord.rs:815:14
version: 0.26.1-dev (c5e09ac9158ec6bde448dd3f328a9bdc95e9dad4)
backtrace:
   0: materialized::handle_panic
             at src/materialized/src/bin/materialized/main.rs:849:21
   1: core::ops::function::Fn::call
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:70:5
   2: std::panicking::rust_panic_with_hook
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:702:17
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:588:13
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:138:18
   5: rust_begin_unwind
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
   6: core::panicking::panic_fmt
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
   7: core::result::unwrap_failed
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/result.rs:1785:5
   8: core::result::Result<T,E>::unwrap
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/result.rs:1078:23
      mz_coord::coord::Coordinator<S>::advance_local_inputs::{{closure}}
             at src/coord/src/coord.rs:811:9
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
      mz_coord::coord::Coordinator<S>::serve::{{closure}}
             at src/coord/src/coord.rs:795:53
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
      tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.2/src/park/thread.rs:263:54
      tokio::coop::with_budget::{{closure}}
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.2/src/coop.rs:102:9
      std::thread::local::LocalKey<T>::try_with
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/local.rs:442:16
      std::thread::local::LocalKey<T>::with
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/local.rs:418:9
  10: tokio::coop::with_budget
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.2/src/coop.rs:95:5
      tokio::coop::budget
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.2/src/coop.rs:72:5
      tokio::park::thread::CachedParkThread::block_on
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.2/src/park/thread.rs:263:31
  11: tokio::runtime::enter::Enter::block_on
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.2/src/runtime/enter.rs:151:13
      tokio::runtime::handle::Handle::block_on
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.18.2/src/runtime/handle.rs:299:9
  12: mz_coord::coord::serve::{{closure}}::{{closure}}
             at src/coord/src/coord.rs:4827:17
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:122:18
  13: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/mod.rs:498:17
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panic/unwind_safe.rs:271:9
      std::panicking::try::do_call
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492:40
      std::panicking::try
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456:19
      std::panic::catch_unwind
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137:14
      std::thread::Builder::spawn_unchecked_::{{closure}}
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/mod.rs:497:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227:5
  14: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/alloc/src/boxed.rs:1861:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/alloc/src/boxed.rs:1861:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys/unix/thread.rs:108:17
  15: <unknown>
  16: <unknown>
```

**Comment by sploiselle** [comment]
From @benesch:

> So the coordinator is not presently expected to tolerate a storaged crash. Is it right that it is surprising that storaged crashed in the first place?

@philip-stoev 's response to this was:

> Ok understood. The storaged oomed due to a leak that was filed separately.

@philip-stoev Do you feel comfortable closing this, or is this an outstanding issue from your perspective?

**Comment by philip-stoev** [comment]
Yes please proceed as you see fit.

**Comment by benesch** [comment]
Will be addressed by #12857.
